pkgs/build-support/rust/rustc-wrapper/default.nix

```diff
@@ -0,0 +1,30 @@
+{ lib, runCommand, rustc, sysroot ? null }:
+
+runCommand "${rustc.pname}-wrapper-${rustc.version}" {
+  preferLocalBuild = true;
+  strictDeps = true;
+  inherit (rustc) outputs;
+
+  env = {
+    prog = "${rustc}/bin/rustc";
+    sysroot = lib.optionalString (sysroot != null) "--sysroot ${sysroot}";
+  };
+
+  passthru = {
+    inherit (rustc) pname version src llvm llvmPackages;
+    unwrapped = rustc;
+  };
+
+  meta = rustc.meta // {
+    description = "${rustc.meta.description} (wrapper script)";
+    priority = 10;
+  };
+} ''
+  mkdir -p $out/bin
+  ln -s ${rustc}/bin/* $out/bin
+  rm $out/bin/rustc
+  substituteAll ${./rustc-wrapper.sh} $out/bin/rustc
```

Comment:
I'm a bit worried about a big shell script interpolated inside a Nix file with various conditional bits being harder to read, but I could be convinced.  I'm also nervous about it being _too_ easy to add conditionals.  Often they're unnecessary and are just hacks to skip rebuilds that then never get cleaned up.  @Ericson2314 wdyt?

Reply:
Personally I find bash conditionals to be a hard-to-read footgun.  But maybe it's just me.

Reply:
Another small reason to keep it in a standalone script file is that we can run shellcheck on it, to minimize bash footguns.

Reply:
@alyssais, shellcheck is *the problem*, not the solution:

https://github.com/NixOS/nixpkgs/blob/5d08323dda0c24cd49594d3ff8bac9810c3620be/pkgs/build-support/cc-wrapper/cc-wrapper.sh#L11-L15 

In addition, `nixpkgs/*/foo.sh` encourage people to stash important data in `nix-support/`, which makes these data inaccessible to Nix expressions -- you can pass things from Nix to bash, but you can't pass things from bash to Nix.

As soon as some upstream dependency starts putting its data in bashworld instead of Nixworld, it forces all of its *downstream* dependencies to do the same thing.  `cc-wrapper` is the worst offender:

https://github.com/NixOS/nixpkgs/blob/5d08323dda0c24cd49594d3ff8bac9810c3620be/pkgs/build-support/cc-wrapper/add-flags.sh#L58-L60

This is why we have such a mess of "hooks" all over the place.